chain/chain/src/store.rs

```diff
@@ -1745,111 +1780,250 @@ impl<'a> ChainStoreUpdate<'a> {
             .insert((height, shard_id), chunk_hash);
     }
 
-    pub fn clear_old_data_on_height(&mut self, height: BlockHeight) -> Result<(), Error> {
+    // This method may be completely replaced in the next updates
+    pub fn build_refcount_map(&mut self) -> Result<(), Error> {
+        let head = self.head()?;
+        let tail = self.tail()?;
+        self.blocks_refcount = HashMap::new();
+        // All heights from tail including head.height
+        for height in tail..=head.height {
+            let blocks_current_height = match self.get_all_block_hashes_by_height(height) {
+                Ok(blocks_current_height) => {
+                    blocks_current_height.values().flatten().cloned().collect()
+                }
+                _ => vec![],
+            };
+            for block_hash in blocks_current_height.iter() {
+                if let Ok(prev_hash) =
+                    self.get_block(&block_hash).map(|block| block.header.prev_hash)
+                {
+                    *self.blocks_refcount.entry(prev_hash).or_insert(0) += 1;
+                }
+            }
+        }
+        Ok(())
+    }
+
+    pub fn get_block_refcount(&self, block_hash: &CryptoHash) -> u64 {
+        self.blocks_refcount.get(block_hash).unwrap_or(&0).clone()
+    }
+
+    pub fn update_tail(&mut self, height: BlockHeight) {
+        self.tail = Some(height);
+    }
+
+    // Clearing block data of `block_hash`, if on a fork.
+    // Clearing block data of `block_hash.prev`, if on the Canonical Chain.
+    pub fn clear_block_data(
+        &mut self,
+        trie: Arc<Trie>,
+        mut block_hash: CryptoHash,
+        is_fork: bool,
+    ) -> Result<(), Error> {
+        let header = match self.get_block_header(&block_hash) {
+            Ok(header) => header.clone(),
+            Err(_) => {
+                // Block data is already cleared
+                panic!("should never happen");
+            }
+        };
+        assert_ne!(header.inner_lite.height, self.get_genesis_height());
+
         let mut store_update = self.store().store_update();
+
+        // 1. Apply revert insertions or deletions from ColTrieChanges for Trie
+        if is_fork {
+            // If the block is on a fork, we delete the state that's the result of applying this block
+            self.store()
+                .get_ser(ColTrieChanges, block_hash.as_ref())?
```

Comment:
Doesn't simple reference (`&block_hash`) work here? It seems that you can clean up quite a few of those explicit `.as_ref()` here.

Reply:
<img width="892" alt="Screen Shot 2020-03-18 at 3 42 05 PM" src="https://user-images.githubusercontent.com/5685493/76952514-09fa4f00-692f-11ea-9907-1a26405f0b22.png">

Reply:
Oh, well, `get_ser` (and `delete`) expects `&[u8]` explicitly. We have `AsRef<[u8]>` in some places, but not there.